include/circt/Dialect/FIRRTL/FIRRTLTypes.td

```diff
@@ -54,6 +54,63 @@ def PassiveType : DialectType<FIRRTLDialect,
   CPred<"$_self.isa<FIRRTLType>() && $_self.cast<FIRRTLType>().isPassive()">,
   "a passive type (contain no flips)", "::circt::firrtl::FIRRTLType">;
 
+def CMemoryType : TypeDef<FIRRTLDialect, "CMemory"> {
+  let mnemonic = "cmemory";
+  let summary = "a behavioral memory";
+  let description = [{
+    Syntax:
+    ```
+    cmemory-type ::= `cmemory` `<` element-type, element-count `>`
+    ```
+
+    The value of a `cmemory` type represents a behavioral memory with unknown
+    ports.  This is produced by `combmem` and `seqmem` declarations and used by
+    `memoryport` declarations to define memories and their ports. A CMemory is
+    similar to a vector of passive element types.
+
+    Examples:
+
+    ```mlir
+    !firrtl.cmemory<uint<32>, 16>
+    !firrtl.cmemory<bundle<a : uint<1>>, 16>
+    ```
+  }];
+
+  let parameters = (ins "FIRRTLType":$elementType, "unsigned":$numElements);
```

Comment:
> From @lattner
> is it possible to put [the verifier code] into the definition of CMemoryType in tblgen, e.g. by changing the element type in let parameters = ... to PassiveType?

I tried this and I'm getting:
```
FIRRTLTypes.td:53:1: error: Record `PassiveType' does not have a field named `cppType'!
```

I'm trying to make some sense of how this works:  `PassiveType` is actually `TypeConstraint` and not a `TypeDef`, and it doesn't seem like we can use `TypeConstraint`s here.  What I am using now `"FIRRTLType"` (with quotes) is actually a short form for `AttrOrTypeParameter<"C++Class">)` which is ok. I don't think there is support to do what we want to do yet.

Reply:
Ok!